File: sdk/python/dc1/resources/jobs.py
"""Jobs resource."""
from __future__ import annotations
import time
from typing import Any, Optional

from ..models import Job
from ..exceptions import APIError, JobTimeoutError

TERMINAL_STATUSES = {'completed', 'failed', 'cancelled'}


class JobsResource:
    def __init__(self, http):
        self._http = http

    def submit(
        self,
        job_type: str,
        params: dict,
        *,
        provider_id: int,
        duration_minutes: float,
        priority: int = 2,
    ) -> Job:
        """Submit a compute job.

        Args:
            job_type: One of 'llm_inference', 'image_generation', 'vllm_serve',
                      'rendering', 'training', 'benchmark', 'custom_container'.
            params: Job-type-specific parameters dict (e.g. {'prompt': '...', 'model': 'llama3'}).
            provider_id: ID of the provider to run on. Use client.providers.list() to discover.
            duration_minutes: Maximum runtime in minutes. Billing is based on actual usage.
            priority: 1=high, 2=normal (default), 3=low.

        Returns:
            Job object with id and initial status.
        """
        body = {
            'provider_id': provider_id,
            'job_type': job_type,
            'duration_minutes': duration_minutes,
            'priority': priority,
            'params': params,
        }
        data = self._http.post('/api/jobs/submit', body)
        submitted_job = data.get('job') if isinstance(data.get('job'), dict) else data
        job_id = str(submitted_job.get('job_id', submitted_job.get('id', data.get('job_id', data.get('id', '')))))
        if not job_id:
            raise APIError('Job submission succeeded but no job_id was returned', status_code=500, response=data)
        return self.get(job_id)

    def get(self, job_id: str) -> Job:
        """Fetch current status and result of a job.

        Args:
            job_id: The job ID returned by submit().

        Returns:
            Job object with current status and result (if completed).
        """
        try:
            data = self._http.get(f'/api/jobs/{job_id}/output')
        except APIError as exc:
            if exc.status_code == 410 and isinstance(exc.response, dict):
                return Job.from_api(exc.response)
            raise
        return Job.from_api(data)

    def wait(
        self,
        job_id: str,
        *,
        timeout: int = 300,
        poll_interval: int = 5,
    ) -> Job:
        """Block until job completes (or timeout).

        Args:
            job_id: The job ID to wait on.
            timeout: Maximum seconds to wait (default 300).
            poll_interval: Seconds between status polls (default 5).

        Returns:
            Completed Job object.

        Raises:
            JobTimeoutError: If the job doesn't complete within timeout seconds.
        """
        deadline = time.monotonic() + timeout
        while True:
            job = self.get(job_id)
            if job.is_done:
                return job
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                raise JobTimeoutError(job_id, timeout)
            time.sleep(min(poll_interval, remaining))

    def list(self, limit: int = 20) -> list[Job]:
        """List recent jobs for the authenticated renter.

        Args:
            limit: Maximum number of jobs to return (default 20).

        Returns:
            List of Job objects, newest first.
        """
        data = self._http.get('/api/jobs/history', params={'limit': limit})
        jobs_raw = data if isinstance(data, list) else data.get('jobs', [])
        return [Job.from_api(j) for j in jobs_raw]

    def cancel(self, job_id: str) -> dict:
        """Cancel a queued or running job. Refunds the estimated cost to the renter's balance.

        Args:
            job_id: The job ID to cancel.

        Returns:
            Dict with ``success`` boolean and ``job_id``.
        """
        data = self._http.post(f'/api/jobs/{job_id}/cancel')
        return {
            'success': bool(data.get('success', False)),
            'job_id': str(data.get('job', {}).get('job_id', job_id) if isinstance(data.get('job'), dict) else job_id),
        }

    def get_logs(self, job_id: str, *, since: Optional[int] = None, limit: int = 200) -> list[dict]:
        """Fetch execution log lines for a job.

        Args:
            job_id: The job ID.
            since: Only return lines after this line number (for incremental tailing).
            limit: Maximum number of lines to return (default 200, max 1000).

        Returns:
            List of log entry dicts with ``line_no``, ``level``, ``message``, and ``logged_at``.
        """
        params: dict = {}
        if since is not None:
            params['since'] = str(since)
        if limit != 200:
            params['limit'] = str(limit)

        data = self._http.get(f'/api/jobs/{job_id}/logs', params=params if params else None)
        raw_logs = data.get('logs', []) if isinstance(data, dict) else []
        return [
            {
                'line_no': l.get('line_no', 0),
                'level': l.get('level', 'info'),
                'message': l.get('message', ''),
                'logged_at': l.get('logged_at', ''),
            }
            for l in raw_logs
        ]

File: sdk/node/src/resources/jobs.ts
import { HttpClient } from '../http';
import { Job, SubmitJobParams, WaitOptions, JobLog } from '../types';
import { JobTimeoutError } from '../errors';

const TERMINAL = new Set(['completed', 'failed', 'cancelled']);

function parseJob(data: Record<string, unknown>): Job {
  const costHalala = (data.cost_halala as number) ?? 0;
  const rawResult = data.result as Record<string, unknown> | string | null;
  let result: Record<string, unknown> | null = null;
  if (rawResult && typeof rawResult === 'object') {
    result = rawResult;
  } else if (typeof rawResult === 'string') {
    result = { output: rawResult };
  }

  return {
    id: String(data.job_id ?? data.id ?? ''),
    status: (data.status as Job['status']) ?? 'unknown',
    jobType: (data.job_type as Job['jobType']) ?? '',
    providerId: (data.provider_id as number) ?? 0,
    durationMinutes: (data.duration_minutes as number) ?? 0,
    costHalala,
    costSar: costHalala / 100,
    submittedAt: (data.submitted_at as string) ?? '',
    startedAt: (data.started_at as string) ?? null,
    completedAt: (data.completed_at as string) ?? null,
    providerName: (data.provider_name as string) ?? null,
    isDone: TERMINAL.has((data.status as string) ?? ''),
    result,
    resultType: (data.result_type as Job['resultType']) ?? null,
    error: (data.error as string) ?? null,
    executionTimeSec: (data.execution_time_sec as number) ?? null,
  };
}

export class JobsResource {
  constructor(private readonly http: HttpClient) {}

  /**
   * Submit a compute job.
   *
   * @param options - Job configuration
   * @returns Job with initial status
   */
  async submit(options: SubmitJobParams): Promise<Job> {
    const body = {
      provider_id: options.providerId,
      job_type: options.jobType,
      duration_minutes: options.durationMinutes,
      priority: options.priority ?? 2,
      params: options.params,
    };
    const data = await this.http.post<Record<string, unknown>>('/api/jobs/submit', body);
    const jobId = String(data.job_id ?? data.id ?? '');
    return this.get(jobId);
  }

  /**
   * Fetch current status and result of a job.
   */
  async get(jobId: string): Promise<Job> {
    const data = await this.http.get<Record<string, unknown>>(`/api/jobs/${jobId}/output`);
    return parseJob(data);
  }

  /**
   * Wait (async poll) until job reaches a terminal state.
   *
   * @throws {JobTimeoutError} when timeoutMs exceeded
   */
  async wait(jobId: string, options: WaitOptions = {}): Promise<Job> {
    const timeoutMs = options.timeoutMs ?? 300_000;
    const pollIntervalMs = options.pollIntervalMs ?? 5_000;
    const deadline = Date.now() + timeoutMs;

    while (true) {
      const job = await this.get(jobId);
      if (job.isDone) return job;

      const remaining = deadline - Date.now();
      if (remaining <= 0) throw new JobTimeoutError(jobId, timeoutMs);

      await new Promise((r) => setTimeout(r, Math.min(pollIntervalMs, remaining)));
    }
  }

  /**
   * List recent jobs for the authenticated renter.
   * Uses GET /api/jobs/history which returns the renter's job history via x-renter-key header.
   */
  async list(limit = 20): Promise<Job[]> {
    const data = await this.http.get<Record<string, unknown>>('/api/jobs/history', {
      limit: String(limit),
    });
    const raw = (data.jobs as unknown[]) ?? [];
    return (raw as Record<string, unknown>[]).map(parseJob);
  }

  /**
   * Cancel a queued or running job. Refunds the estimated cost to the renter's balance.
   *
   * @param jobId - The job ID (e.g. `job-1234-abcdef`)
   */
  async cancel(jobId: string): Promise<{ success: boolean; jobId: string }> {
    const data = await this.http.post<Record<string, unknown>>(`/api/jobs/${jobId}/cancel`);
    return {
      success: (data.success as boolean) ?? false,
      jobId: String((data.job as Record<string, unknown>)?.job_id ?? jobId),
    };
  }

  /**
   * Fetch execution log lines for a job.
   *
   * @param jobId - The job ID
   * @param options.since - Only return lines after this line number (for incremental tailing)
   * @param options.limit - Max lines to return (default 200, max 1000)
   */
  async getLogs(jobId: string, options: { since?: number; limit?: number } = {}): Promise<JobLog[]> {
    const params: Record<string, string> = {};
    if (options.since != null) params.since = String(options.since);
    if (options.limit != null) params.limit = String(options.limit);

    const data = await this.http.get<Record<string, unknown>>(
      `/api/jobs/${jobId}/logs`,
      params,
    );
    const raw = (data.logs as Record<string, unknown>[]) ?? [];
    return raw.map((l) => ({
      lineNo: (l.line_no as number) ?? 0,
      level: (l.level as JobLog['level']) ?? 'info',
      message: (l.message as string) ?? '',
      loggedAt: (l.logged_at as string) ?? '',
    }));
  }
}
